go/tasks/plugins/presto/execution_state.go

```diff
@@ -0,0 +1,528 @@
+			break
+		}
+	}
+	logger.Infof(ctx, "Created a resource constraints spec: [%v]", constraintsSpec)
+	return constraintsSpec
+}
+
+func GetNextQuery(
+	ctx context.Context,
+	tCtx core.TaskExecutionContext,
+	currentState ExecutionState) (Query, error) {
+
+	switch currentState.QueryCount {
+	case 0:
+		prestoCfg := config.GetPrestoConfig()
+		tempTableName := generateRandomString(32)
+		routingGroup, catalog, schema, statement, err := GetQueryInfo(ctx, tCtx)
+		if err != nil {
+			return Query{}, err
+		}
+
+		statement = fmt.Sprintf(`CREATE TABLE hive.flyte_temporary_tables.%s_temp AS %s`, tempTableName, statement)
+
+		prestoQuery := Query{
+			Statement: statement,
+			ExtraArgs: client.PrestoExecuteArgs{
+				RoutingGroup: resolveRoutingGroup(ctx, routingGroup, prestoCfg),
+				Catalog:      catalog,
+				Schema:       schema,
+			},
+			TempTableName:     tempTableName + "_temp",
+			ExternalTableName: tempTableName + "_external",
+		}
+
+		return prestoQuery, nil
+
+	case 1:
+		cfg := config.GetPrestoConfig()
+		externalLocation := getExternalLocation(cfg.AwsS3ShardFormatter, cfg.AwsS3ShardCount)
+
+		statement := fmt.Sprintf(`
+CREATE TABLE hive.flyte_temporary_tables.%s (LIKE hive.flyte_temporary_tables.%s)
+WITH (format = 'PARQUET', external_location = '%s')`,
+			currentState.CurrentPrestoQuery.ExternalTableName,
+			currentState.CurrentPrestoQuery.TempTableName,
+			externalLocation,
+		)
+		currentState.CurrentPrestoQuery.Statement = statement
+		return currentState.CurrentPrestoQuery, nil
+
+	case 2:
+		statement := `
+INSERT INTO hive.flyte_temporary_tables.%s
+SELECT *
+FROM hive.flyte_temporary_tables.%s`
+		statement = fmt.Sprintf(statement, currentState.CurrentPrestoQuery.ExternalTableName, currentState.CurrentPrestoQuery.TempTableName)
+		currentState.CurrentPrestoQuery.Statement = statement
+		return currentState.CurrentPrestoQuery, nil
+
+	case 3:
+		statement := fmt.Sprintf(`DROP TABLE %s`, currentState.CurrentPrestoQuery.TempTableName)
+		currentState.CurrentPrestoQuery.Statement = statement
+		return currentState.CurrentPrestoQuery, nil
+
+	case 4:
+		statement := fmt.Sprintf(`DROP TABLE %s`, currentState.CurrentPrestoQuery.ExternalTableName)
+		currentState.CurrentPrestoQuery.Statement = statement
+		return currentState.CurrentPrestoQuery, nil
+
+	default:
+		return currentState.CurrentPrestoQuery, nil
+	}
+}
+
+func generateRandomString(length int) string {
+	const letters = "0123456789abcdefghijklmnopqrstuvwxyz"
+	bytes, err := generateRandomBytes(length)
+	if err != nil {
+		return ""
+	}
+	for i, b := range bytes {
+		bytes[i] = letters[b%byte(len(letters))]
+	}
+	return string(bytes)
+}
+
+func generateRandomBytes(length int) ([]byte, error) {
+	b := make([]byte, length)
+	_, err := rand.Read(b)
+	// Note that err == nil only if we read len(b) bytes.
+	if err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+func getExternalLocation(shardFormatter string, shardLength int) string {
```

Comment:
Can you move the whole sharding thing outside of the presto plugin? maybe in some storage util in plugins/core ?

Reply:
This sharding logic currently only exists in flytekit for Hive I believe. Since currently, only the Presto plugin uses this, can we keep it together with the rest of the Presto logic?